compiler/rustc_codegen_llvm/src/back/archive.rs

```diff
@@ -332,22 +406,46 @@ impl<'a> LlvmArchiveBuilder<'a> {
         }
     }
 
-    fn i686_decorated_name(import: &DllImport) -> CString {
+    fn i686_decorated_name(import: &DllImport, mingw: bool) -> String {
         let name = import.name;
-        // We verified during construction that `name` does not contain any NULL characters, so the
-        // conversion to CString is guaranteed to succeed.
-        CString::new(match import.calling_convention {
-            DllCallingConvention::C => format!("_{}", name),
-            DllCallingConvention::Stdcall(arg_list_size) => format!("_{}@{}", name, arg_list_size),
+        let prefix = if mingw { "" } else { "_" };
+
+        match import.calling_convention {
+            DllCallingConvention::C => format!("{}{}", prefix, name),
+            DllCallingConvention::Stdcall(arg_list_size) => {
+                format!("{}{}@{}", prefix, name, arg_list_size)
+            }
             DllCallingConvention::Fastcall(arg_list_size) => format!("@{}@{}", name, arg_list_size),
             DllCallingConvention::Vectorcall(arg_list_size) => {
                 format!("{}@@{}", name, arg_list_size)
             }
-        })
-        .unwrap()
+        }
     }
 }
 
 fn string_to_io_error(s: String) -> io::Error {
     io::Error::new(io::ErrorKind::Other, format!("bad archive: {}", s))
 }
+
+fn find_binutils_dlltool(sess: &Session) -> OsString {
+    if sess.host.llvm_target != sess.target.llvm_target {
+        assert!(sess.target.options.is_like_windows && !sess.target.options.is_like_msvc);
+        let prefixed_dlltool = if sess.target.arch == "x86" {
+            "i686-w64-mingw32-dlltool"
+        } else {
+            "x86_64-w64-mingw32-dlltool"
+        };
+        let prefixed_dlltool = if cfg!(windows) {
+            [prefixed_dlltool, "exe"].concat()
+        } else {
+            prefixed_dlltool.to_string()
+        };
+        for dir in env::split_paths(&env::var_os("PATH").unwrap_or_default()) {
+            let full_path = dir.join(&prefixed_dlltool);
+            if full_path.is_file() {
+                return full_path.into_os_string();
+            }
+        }
```

Comment:
Porting @nagisa's comment from https://github.com/rust-lang/rust/pull/88801#discussion_r711034769:
> This should also either verify that the binary is executable at all in some way or provide a way to override this detection. Right now if the system is borked enough to have a non-working i686-w64-mingw32-dlltool but a working dlltool, then there is fairly little the user user can do to work around the problem.

I agree that it would be good to allow the user to force the compiler to use a specific `dlltool.exe` in some way. I don't know what the best strategy is though.

I also don't understand the requirements here clearly. When do we need to use `i686-w64-mingw32-dlltool.exe`, when do we need to use `x86_64-w64-mingw32-dlltool.exe`, and when should we use `dlltool.exe`?

Reply:
When cross-compiling `i686...` is for 32-bit targets, `x86_64...` is for 64-bit targets, plain `dlltool.exe` is used for native compilation (target == host).

Reply:
Does the addition of `-Z dlltool` satisfy this comment?  Or is further action required?